python/indexing/CMakeLists.txt

```diff
@@ -0,0 +1,64 @@
+include(AddMLIRPython)
+mlir_configure_python_dev_packages()
+mlir_detect_pybind11_install()
+
+add_compile_definitions("MLIR_PYTHON_PACKAGE_PREFIX=mlir_indexing.")
+
+# find the directory where the python bindings IRModule.h lives
+# and add it as an include directory in order to extend/play with the bindings
+get_target_property(mlir_python_sources MLIRPythonExtension.Core INTERFACE_SOURCES)
+list(GET mlir_python_sources 0 HEADER)
+get_filename_component(header_dir ${HEADER} DIRECTORY)
+string(REPLACE "$<BUILD_INTERFACE:" "" header_dir_clean "${header_dir}")
+set(header_dir_interface "${header_dir}>")
+
+execute_process(COMMAND python ${CMAKE_CURRENT_SOURCE_DIR}/regex_mlir_bindings.py ${header_dir_clean}/IRModule.h)
+execute_process(COMMAND python ${CMAKE_CURRENT_SOURCE_DIR}/regex_mlir_bindings.py ${header_dir_clean}/IRCore.cpp)
+execute_process(COMMAND python ${CMAKE_CURRENT_SOURCE_DIR}/regex_mlir_bindings.py ${header_dir_clean}/IRAffine.cpp)
```

Comment:
The reason for all this wackiness is because you want to use the template classes `PyConcreteType`, `PyConcreteValue`, `PyConcreteAttribute`, and etc. in [IRCore.cpp](https://github.com/llvm/llvm-project/blob/0ae342f45bedd29e34690de087011a9da4db6a65/mlir/lib/Bindings/Python/IRCore.cpp#L1974). Right now that can't be done for two reasons:

1. The symbols aren't exported in C++ (i.e., needs `PYBIND11_EXPORT __attribute__((visibility("default")))` and so they're not visible at runtime (when you load `_mlir.cpython.so`);
2. The C++ classes are bound (using pybind) with `py::module_local()` annotation which prevents them from being subclassed in Python (that annotation has the effect of hiding the constructor in Python);
3. The headers in `mlir/lib/Bindings/Python` are not in an `include/` directory.

Overall very annoying and cumbersome to hackaround, but the hacks work - I've been running them for [months now](https://github.com/makslevental/nelli/actions). Like I mentioned in the PR description, longterm (I'm hoping) they'll go away.

Reply:
These symbols are intentionally unexported. For attributes and types, we should use `mlir_attr/type_subclass` which are exported. Nobody just cared enough to update the built-in types from `PyConcrete*` to these.

For values, I think https://reviews.llvm.org/D147758 adds a similar mechanism so we won't this dance anymore. Or is there something else?

Reply:
Yea I think this is valid - I had in mind possibly some other extension points but no need to frontload them.